main/src/main/scala/sbt/internal/nio/CheckBuildSources.scala

```diff
@@ -59,7 +58,7 @@ private[sbt] object CheckBuildSources {
   }
   private[sbt] def buildSourceFileInputs: Def.Initialize[Seq[Glob]] = Def.setting {
     if (onChangedBuildSource.value != IgnoreSourceChanges) {
-      val baseDir = (LocalRootProject / baseDirectory).value
+      val baseDir = baseDirectory.value
       val sourceFilter = "*.{sbt,scala,java}"
       val projectDir = baseDir / "project"
```

Comment:
Wouldn't this end up tracking `sub/project/*.scala`? Maybe metabuild source inputs needs to be split up, and added back at global level.

Reply:
Yeah, that's intentional. In the project directory, top level scala files are allowed in compile.

Reply:
It's common for builds to have something like project/Dependencies.scala and then refer to object Dependencies in the top level build.sbt.

Reply:
I understand you need to track `project/*.scala`. There are two axes to traverse. One is subprojects, the other is metabuild. Currently those two concerns are mixed, so you end up with `sub/project/*.scala`.

Reply:
Right, but why can't my subproject also have a metabuild?

Reply:
A metabuild can have subprojects (, which I believe this code still doesn't track), but a metabuild of a subproject doesn't make sense since subprojects is a member of the proper build, and only a build can have its meta-build.

Reply:
I don't know the answer, but that appears to be true. I guess I misunderstood how recursion works with builds.

Reply:
To put it into examples:

```
project/D.scala                // track
project/plugins.sbt            // track
project/foo/A.scala            // track (bonus points)
project/src/main/scala/B.scala // track
project/project/C.scala        // track
project/project/plugins.sbt    // track
build.sbt                      // track
sub/build.sbt                  // track
sub/project/E.scala            // do not track
```

Reply:
I can't get the bonus points without degrading performance because searching for all of the scala files would require listing the target directory which can be extremely slow. Otherwise, I think that #5086 now handles the other cases.